srml/offences/src/lib.rs

```diff
@@ -159,8 +161,111 @@ impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T
 			.collect::<Vec<_>>();
 
 		T::OnOffenceHandler::on_offence(
-			&all_offenders,
+			&concurrent_offenders,
 			&slash_perbill,
 		);
+
+		cache.save()
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Compute the ID for the given report properties.
+	///
+	/// The report id depends on the offence kind, time slot and the id of offender.
+	fn report_id<O: Offence<T::IdentificationTuple>>(
+		time_slot: &O::TimeSlot,
+		offender: &T::IdentificationTuple,
+	) -> ReportIdOf<T> {
+		(O::ID, time_slot.encode(), offender).using_encoded(T::Hashing::hash)
+	}
+
+	/// Goes through all offenders in a report and record each unique offence updating indexes.
+	///
+	/// Returns `true` if there were unique reports.
+	fn collect_unique_reports<O: Offence<T::IdentificationTuple>>(
+		reporters: Vec<T::AccountId>,
+		time_slot: &O::TimeSlot,
+		offenders: Vec<T::IdentificationTuple>,
+		index_cache: &mut ReportIndexCache<T, O>,
+	) -> BTreeSet<T::IdentificationTuple> {
+		let mut new_offenders = BTreeSet::new();
+
+		for offender in offenders {
+			let report_id = Self::report_id::<O>(time_slot, &offender);
+
+			if !<Reports<T>>::exists(&report_id) {
+				new_offenders.insert(offender.clone());
+				<Reports<T>>::insert(
+					&report_id,
+					OffenceDetails {
+						offender,
+						reporters: reporters.clone(),
+					},
+				);
+
+				index_cache.insert(time_slot, report_id);
+			}
+		}
+
+		new_offenders
+	}
+}
+
+/// An auxilary struct for dealing with caches of report indexes localized for a specific offence
+/// kind.
+///
+/// It is expected that access to the indexes will be only performed through this struct while it
+/// exists.
+struct ReportIndexCache<T: Trait, O: Offence<T::IdentificationTuple>> {
```

Comment:
I'm not convinced to call it a `cache`, the struct is strictly responsible for:
1. Processing new reports (new offenders)
2. Providing the list of new_offenders and all offenders of the same time slot (`concurrent_reports`)
3. Persisting the list at the end.

I'd merge this logic with `collect_*` thing, and just do all alterations here, while only keeping the offence reporting (event) and perbill slashing in the main part.